File: src/quicksort.clj
(defn qsort [L]
  (if (empty? L)
      '()
      (let [[pivot & L2] L]
           (lazy-cat (qsort (for [y L2 :when (<  y pivot)] y))
                     (list pivot)
                     (qsort (for [y L2 :when (>= y pivot)] y))))))

(use 'clojure.test)

(deftest output-list-is-sorted
  (let [input '(4, 3, 5, 1, 6, 1, 3)
        expected '(1, 1, 3, 3, 4, 5, 6)]
    (is (= expected (qsort input)))
    )
  )

(run-all-tests)

File: src/quicksort.py
def qsort(list):
    if not list:
        return []
    else:
        pivot = list[0]
        less = [x for x in list     if x <  pivot]
        more = [x for x in list[1:] if x >= pivot]
        return qsort(less) + [pivot] + qsort(more)

input_list = [3, 5, 4, 5, 2, 1, 2]
expected_list = [1, 2, 2, 3, 4, 5, 5]

print(expected_list == qsort(input_list))
